ipam/ipam.go

```diff
@@ -0,0 +1,250 @@
+	}
+	if config.Provider == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Provider must not be empty", config)
+	}
+	if config.CommonDomain == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CommonDomain must not be empty", config)
+	}
+	if config.HostClusterName == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.HostClusterName must not be empty", config)
+	}
+
+	s := &IPAM{
+		hostFramework: config.HostFramework,
+		logger:        config.Logger,
+		provider:      config.Provider,
+
+		commonDomain:    config.CommonDomain,
+		hostClusterName: config.HostClusterName,
+	}
+
+	return s, nil
+}
+
+func (c *IPAM) Test(ctx context.Context) error {
+
+	var (
+		// Clusters to be created during this test.
+		clusterOne   = c.hostClusterName + "-cluster0"
+		clusterTwo   = c.hostClusterName + "-cluster1"
+		clusterThree = c.hostClusterName + "-cluster2"
+		clusterFour  = c.hostClusterName + "-cluster3"
+
+		// allocatedSubnets[clusterName]subnetCIDRStr
+		allocatedSubnets = make(map[string]string)
+		err              error
+		// guestFrameworks[clusterName]guestFramework
+		guestFrameworks = make(map[string]*framework.Guest)
+	)
+
+	defer func() {
+		for _, cn := range []string{clusterOne, clusterTwo, clusterThree, clusterFour} {
+			err := c.provider.DeleteCluster(cn)
+			if err != nil {
+				c.logger.LogCtx(ctx, "level", "error", "message", fmt.Sprintf("cluster %s deletion failed: %#v", cn, err))
+			}
+		}
+	}()
+
+	// This is a list of clusters that are created in first test phase.
+	clusters := []string{clusterOne, clusterTwo, clusterThree}
+
+	{
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating three guest clusters: %#v", clusters))
+
+		for _, cn := range clusters {
+			err = c.provider.CreateCluster(cn)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("created guest cluster %s", cn))
+		}
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("waiting for three guest clusters to be ready: %#v", clusters))
+
+		for _, cn := range clusters {
+			cfg := framework.GuestConfig{
+				Logger: c.logger,
+
+				ClusterID:    cn,
+				CommonDomain: c.commonDomain,
+			}
+			guestFramework, err := framework.NewGuest(cfg)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			guestFrameworks[cn] = guestFramework
+			err = guestFramework.Setup()
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("guest cluster %s ready", cn))
+		}
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("all three guest clusters are ready: %#v", clusters))
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", "verifying that subnet allocations don't overlap")
+
+		for _, cn := range clusters {
+			awsConfig, err := c.hostFramework.AWSCluster(cn)
```

Comment:
This looks provider specific.

Reply:
Yes, this is. I made a note somewhere earlier that this first implementation is indeed provider specific. I'll refactor this when there's another provider taking this into use. That way I know what's actually needed and can create better abstractions.